osquery/dispatcher/posix/query_profiler.cpp

```diff
@@ -0,0 +1,162 @@
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+          std::chrono::seconds(end_stat.tv_sec) +
+          std::chrono::microseconds(end_stat.tv_usec))
+          .count(),
+      stat_name + ".millis");
+}
+
+void recordRusageStatDifference(const struct rusage& start_stats,
+                                const struct rusage& end_stats,
+                                const std::string& monitoring_path_prefix) {
+  recordRusageStatDifference(
+      0, end_stats.ru_maxrss, monitoring_path_prefix + ".rss.max.kb");
+
+  recordRusageStatDifference(start_stats.ru_maxrss,
+                             end_stats.ru_maxrss,
+                             monitoring_path_prefix + ".rss.increase.kb");
+
+  recordRusageStatDifference(start_stats.ru_inblock,
+                             end_stats.ru_inblock,
+                             monitoring_path_prefix + ".input.load");
+
+  recordRusageStatDifference(start_stats.ru_oublock,
+                             end_stats.ru_oublock,
+                             monitoring_path_prefix + ".output.load");
+
+  recordRusageStatDifference(start_stats.ru_utime,
+                             end_stats.ru_utime,
+                             monitoring_path_prefix + ".time.user");
+
+  recordRusageStatDifference(start_stats.ru_stime,
+                             end_stats.ru_stime,
+                             monitoring_path_prefix + ".time.system");
+}
+
+enum class RusageError { FatalError = 1 };
+Expected<struct rusage, RusageError> callRusage() {
+  struct rusage stats;
+  const int who = getRusageWho();
+  auto rusage_status = getrusage(who, &stats);
+  if (rusage_status != -1) {
+    return stats;
+  } else {
+    return createError(RusageError::FatalError, "")
+           << "Linux query profiling failed. error code: " << rusage_status
+           << " message: " << boost::io::quoted(strerror(errno));
+  }
+}
+
+void launchQueryWithPosixProfiling(const std::string& name,
+                                   std::function<Status()> launchQuery) {
+  const auto start_time_point = std::chrono::steady_clock::now();
+  auto rusage_start = callRusage();
+
+  if (!rusage_start) {
+    LOG(ERROR) << "rusage_start error: "
+               << rusage_start.getError().getFullMessageRecursive();
+  }
+
+  const auto status = launchQuery();
+  const auto monitoring_path_prefix =
+      (boost::format("scheduler.executing_query.%s.%s") % name %
+       (status.ok() ? "success" : "failure"))
+          .str();
+
+  if (rusage_start) {
+    const auto rusage_end = callRusage();
+
+    if (rusage_end) {
+      recordRusageStatDifference(
+          *rusage_start, *rusage_end, monitoring_path_prefix);
+    } else {
+      LOG(ERROR) << "rusage_end error: "
+                 << rusage_end.getError().getFullMessageRecursive();
+    }
+  }
+
+  const auto query_duration =
+      std::chrono::duration_cast<std::chrono::microseconds>(
+          std::chrono::steady_clock::now() - start_time_point);
```

Comment:
Do you think it is good idea to include the time of monitoring key creating and time of all routine in `recordRusageStatDifference` in `query_duration`? Why?

Reply:
query monitoring is not rusage specific, that's why I don't think it should be somewhere completely separated.